### ISSUE TITLE
{LazyList, immutable.BitSet} builder may continue modifying previously-released LazyLists

### ISSUE THREAD

**Issue by joshlemer** [scala/bug#11160]
```scala
    val b = LazyList.newBuilder[Int]
    b.addOne(1000)
    val r = b.result()
    b.addOne(9) 
    r.toVector // Vector(1000, 9)
```

```scala
    val b = immutable.BitSet.newBuilder
    b.addOne(1000)
    val r = b.result()
    b.addOne(9)
    r // BitSet(9, 1000)
```

**Comment by NthPortal** [comment]
`LazyList` calls `mapResult` off of `ArrayBuffer.newBuilder`, so the problem would seem to lie there

**Comment by joshlemer** [comment]
@Ichoran do you have an opinion on what the behavior here should be? From the existence of `ReusableBuilder` it seems implied that regular builders are not reusable, and so adding elements to Builders after `result()` is a violation of its contract. But then, what is the correct behaviour? Throw exception, or do nothing?

**Comment by xuwei-k** [comment]
- https://github.com/scala/scala/blob/9bfd7d82c277aefa1fe987724f445a5d2d6a53a1/src/library/scala/collection/immutable/BitSet.scala#L91
- https://github.com/scala/scala/blob/9bfd7d82c277aefa1fe987724f445a5d2d6a53a1/src/library/scala/collection/immutable/BitSet.scala#L95-L116

use `fromBitMask` instead of `fromBitMaskNoCopy` ? 🤔

**Comment by szeiger** [comment]
I wonder if we need `ResuableBuilder` at all. `Builder` already has a `clear` method from `Growable`. Is there any reason why you would expect `clear` to work before producing a result but not afterwards (in order to reset the builder to its initial state)? I'll look through the existing builder implementations to get rid of `ReusableBuilder`. The first few non-reusable `Builder` subclasses that I looked at are actually reusable.

**Comment by joshlemer** [comment]
>I wonder if we need ResuableBuilder at all. 

@szeiger this is what I have been wondering for a while, it seems to me that all builders should behave nicely, and in builders I have implemented, I have made sure they are "reusable" even without extending that trait.

**Comment by Ichoran** [comment]
I think the biggest barrier was that some mutable collections were their own builder (e.g. `ArrayBuffer`), which makes reuse tricky without substantial inefficiencies (you at least have to create one new object and transfer all the innards over).

**Comment by szeiger** [comment]
Good point. So we still need non-reusable builders for that. It also answers the question what the semantics after calling `result` should be. Similar to Iterators which can be reused and updated internally, the behavior in general is undefined.
